### PR TITLE
TopPic Proforma Parser

### PR DESCRIPTION
**Pull request by mikehollas123**
Added a parser and tests for converting TopPic Proteoform strings into ProForma

### CODE REVIEW

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace TopDownProteomics.ProForma
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:7]: let's start moving toward file scoped namespaces.

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+{
+    /// <summary>
+    /// A parser for TopPic strings into a ProformaTerm <see cref="ProFormaTerm"/>
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:10]: The name is TopPIC, not TopPic ... not sure if we should change the class name, but comments should case it right

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            var numberMatch = _numberRx.Match(ptmstring);
+
+            if (numberMatch.Success)
+                return new ProFormaDescriptor(ProFormaKey.Mass, numberMatch.Value);
+
+            // Find and throw exception if there is a *
+            if (ptmstring.Contains('*'))
+                throw new Exception("multiple mods are not currently supported");
+
+            if (_modLookup is not null && _modLookup.ContainsKey(ptmstring))
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:167]: could do `if (_modLookup?.ContainsKey(ptmstring) == true)` if you like that better

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                    throw new Exception($"Failed to parse unimod id {splitLine[1]}");
+            }
+
+            return modLookup;
+        }
+
+
+        private string GetFullyStrippedSequence(string sequence) => Regex.Replace(sequence, @"\[[\w_+-.]+\]|[()]", "");
+
+        private Dictionary<int, int> GetIndexLoopup(string sequence)
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:88]: spelling, `GetIndexLookup`

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                        throw new Exception($"invalid unimod id");
+
+                }
+                else
+                    throw new Exception($"Failed to parse unimod id {splitLine[1]}");
+            }
+
+            return modLookup;
+        }
+
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:84]: SPACE! SPACE!

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+                if (Int32.TryParse(splitLine[4], out var uniModNumber))
+                {
+                    if (uniModNumber > 0)
+                        modLookup.Add(name, new ProFormaDescriptor(ProFormaKey.Identifier, ProFormaEvidenceType.Unimod, $"UNIMOD:{uniModNumber}"));
+                    else if (uniModNumber == -1)
+                        modLookup.Add(name, new ProFormaDescriptor(ProFormaKey.None, name)); // maybe could do mass instead?
+                    else
+                        throw new Exception($"invalid unimod id");
+
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:76]: space

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        {
+            IDictionary<string, ProFormaDescriptor> modLookup = new Dictionary<string, ProFormaDescriptor>();
+
+            using StreamReader reader = new StreamReader(modFile);
+
+            while (!reader.EndOfStream)
+            {
+                var line = reader.ReadLine();
+
+                if (line.Length == 0 | line.StartsWith("#"))
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:59]: I would ignore lines with only spaces (so length > 0)

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+                    continue;
+
+                //# To input a modification, use the following format:
+                //# Name,Mass,Residues,Position,UnimodID
+
+                var splitLine = line.Split(',');
+                var name = splitLine[0];
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:66]: Add defensive logic to check the array size == 5

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+                if (Int32.TryParse(splitLine[4], out var uniModNumber))
+                {
+                    if (uniModNumber > 0)
+                        modLookup.Add(name, new ProFormaDescriptor(ProFormaKey.Identifier, ProFormaEvidenceType.Unimod, $"UNIMOD:{uniModNumber}"));
+                    else if (uniModNumber == -1)
+                        modLookup.Add(name, new ProFormaDescriptor(ProFormaKey.None, name)); // maybe could do mass instead?
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:73]: ya, mass might be better or you could do multiple descriptors. Generally, we should consider including an info tag with the name along with the ID or mass.

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        }
+
+        /// <summary>
+        /// Gets the proforma term.
+        /// </summary>
+        /// <param name="sequence">The sequence.</param>
+        /// <returns></returns>
+        public ProFormaTerm ParseTopPicString(string sequence)
+        {
+            //first remove terminaltags if there!
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:43]: spelling

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+            sequence = RemoveTerminalAAs(sequence);
+            var ptms = FindPTMs(sequence);
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:45]: tuples are fine to use, but this line and the following aren't very readable. Switch to something like:
`var (tags, nTerm, cTerm) = FindPTMs(sequence);`

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+                }
+                else
+                    throw new Exception($"Failed to parse unimod id {splitLine[1]}");
+            }
+
+            return modLookup;
+        }
+
+
+        private string GetFullyStrippedSequence(string sequence) => Regex.Replace(sequence, @"\[[\w_+-.]+\]|[()]", "");
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:86]: what is the first clause in the regex trying to do? It seems like there are missing characters that might cause problems ...

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+        private Dictionary<int, int> GetIndexLoopup(string sequence)
+        {
+            Dictionary<int, int> indexLoopup = new Dictionary<int, int>();
+
+            bool inBraket = false;
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:92]: spelling

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+            List<ProFormaDescriptor> C_terms = new List<ProFormaDescriptor>();
+            List<ProFormaTag> Tags = new List<ProFormaTag>();
+
+            foreach (Match match in _modRx.Matches(sequence))
+            {
+                var startIndex = indexLoopup[match.Groups[1].Index];
+                var ptms = match.Groups[2].Captures;
+
+                if (ptms.Count > 1)
+                    throw new Exception("multiple mods are not currently accepeted");
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:123]: spelling

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+
+                if (startIndex == 0 && match.Groups[1].Length == 1)  // check for ambiguoous mods that include the start -> just make tags
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:125]: spelling

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+                }
+            }
+            return indexLoopup;
+        }
+
+        private Tuple<List<ProFormaDescriptor>, List<ProFormaDescriptor>, List<ProFormaTag>> FindPTMs(string sequence)
+        {
+            var indexLoopup = GetIndexLoopup(sequence);
+
+            List<ProFormaDescriptor> N_terms = new List<ProFormaDescriptor>();
```
Comment (rfellers) [inline src/TopDownProteomics/ProForma/TopPicProformaParser.cs:113]: odd variable naming for .NET

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs (hunk 12 above)
Comment (mikehollas123) [reply src/TopDownProteomics/ProForma/TopPicProformaParser.cs:86]: it needs to pull out everything inside []. I tried a couple of things. but you can't have something that is inclusive of ']' or it breaks!

Comment (mikehollas123) [reply src/TopDownProteomics/ProForma/TopPicProformaParser.cs:86]: `\[[^]^[]+\]|[()]` 
This seems to work

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs (hunk 2 above)
Comment (mikehollas123) [reply src/TopDownProteomics/ProForma/TopPicProformaParser.cs]: Left the class name for now

File: src/TopDownProteomics/ProForma/TopPicProformaParser.cs (hunk 9 above)
Comment (mikehollas123) [reply src/TopDownProteomics/ProForma/TopPicProformaParser.cs]: Added info tags and better handling of masses

File: tests/TopDownProteomics.Tests/TopDownProteomics.Tests.csproj [hunk 17]
```diff
@@ -18,6 +18,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <None Remove="TestData\topPicTestMods.txt" />
```
Comment (rfellers) [inline tests/TopDownProteomics.Tests/TopDownProteomics.Tests.csproj:21]: the changes to this file can removed. VS likes to add this for some reason even though it isn't needed.

File: tests/TopDownProteomics.Tests/ProForma/ToPicParserTests.cs [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+    [TestCase("W.(G)[Oxidation]DGCAQKNKPGVYTK(V)[Phospho]YNYVKWIKNTIAANS.", "[UNIMOD:35|Info:Oxidation]-GDGCAQKNKPGVYTKV[UNIMOD:21|Info:Phospho]YNYVKWIKNTIAANS")]
+    [TestCase(".GDGCAQKNKPGVYTK(V)[Phospho]YNYVKWIKNTIAANS.", "GDGCAQKNKPGVYTKV[UNIMOD:21|Info:Phospho]YNYVKWIKNTIAANS")]
+    [TestCase("W.GDGCAQKNKPGVYTKVYNYVKWIKNTIAAN(S)[Phospho].", "GDGCAQKNKPGVYTKVYNYVKWIKNTIAANS-[UNIMOD:21|Info:Phospho]")]
+    [TestCase("W.GDGCAQKNKPGVYTKVYNYVKWIKNTIAANS.", "GDGCAQKNKPGVYTKVYNYVKWIKNTIAANS")]
+    [TestCase(".(G)[Test1]DGCAQKNKPGVYTKVYNYVKWIKNTIAANS.", "[+59.000000|Info:Test1]-GDGCAQKNKPGVYTKVYNYVKWIKNTIAANS")]
+    [TestCase("W.(G)[T@s!1]DGCAQKNKPGVYTKVYNYVKWIKNTIAANS.", "[T@s!1]-GDGCAQKNKPGVYTKVYNYVKWIKNTIAANS")]
+    [TestCase("W.(G)[Test_2]DGCAQKNKPGVYTKVYNYVKWIKNTIAANS.", "[+59.000000|Info:Test_2]-GDGCAQKNKPGVYTKVYNYVKWIKNTIAANS")]
+    [TestCase(".(G)[Ox_plus1]DGCAQKNKPGVYTKVYNYVKWIKNTIAANS.", "[+17.123000|Info:Ox_plus1]-GDGCAQKNKPGVYTKVYNYVKWIKNTIAANS")]
+    [TestCase(".(G)[+23.9987]DGCAQKNKPGVYTKVYNYVKWIKNTIAANS.", "[+23.9987]-GDGCAQKNKPGVYTKVYNYVKWIKNTIAANS")]
+
```
Comment (rfellers) [inline tests/TopDownProteomics.Tests/ProForma/ToPicParserTests.cs:28]: space :(

File: tests/TopDownProteomics.Tests/ProForma/ToPicParserTests.cs [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+
+    /// <summary>
+    /// Tests the TopPic Proforma Parser with no mod file.
+    /// </summary>
+    [Test]
+    [TestCase("M.A(AAA)[Phospho]AAA.C", "A(AAA)[Phospho]AAA")]
+    [TestCase("W.(G)[Oxidation]DGCAQKNKPGVYTK(V)[Phospho]YNYVKWIKNTIAANS.", "[Oxidation]-GDGCAQKNKPGVYTKV[Phospho]YNYVKWIKNTIAANS")]
+    [TestCase("W.(G)[asdf4fdfsd6!]DGCAQKNKPGVYTKYNYVKWIKNTIAANS.", "[asdf4fdfsd6!]-GDGCAQKNKPGVYTKYNYVKWIKNTIAANS")]
+
```
Comment (rfellers) [inline tests/TopDownProteomics.Tests/ProForma/ToPicParserTests.cs:47]: space :(

File: tests/TopDownProteomics.Tests/ProForma/ToPicParserTests.cs [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+    public void CompareToProForma(string topPIC, string proForma)
+    {
+        var topicParser = new TopPicProformaParser(@".\TestData\topPicTestMods.txt");
+        var term = topicParser.ParseTopPicString(topPIC);
+
+        var writer = new ProFormaWriter();
+
+        Assert.AreEqual(proForma, writer.WriteString(term));
+    }
+
```
Comment (rfellers) [inline tests/TopDownProteomics.Tests/ProForma/ToPicParserTests.cs:38]: space

**Review by rfellers — CHANGES_REQUESTED** [review]
There are 4 code coverage warnings that need unit tests that trigger the exceptions. Also, some extra spaces that make anal-retentive me itchy.

**Review by rfellers — APPROVED** [review]

**Pull request merged**